### PR TITLE
Blocking harmful methods in spel in runtime

### PR DESCRIPTION
**Pull request by maciej-brzezinski**

### CODE REVIEW

File: engine/benchmarks/src/test/scala/pl/touk/nussknacker/engine/benchmarks/spel/SpelSecurityBlacklistBenchmark.scala [hunk 1]
```diff
@@ -0,0 +1,67 @@
...
+  @BenchmarkMode(Array(Mode.AverageTime))
+  @OutputTimeUnit(TimeUnit.MICROSECONDS)
+  def benchmark(): AnyRef = {
+    setup.test()
+  }
+}
+
+
+
+/**
```
Comment (dswiecki) [inline engine/benchmarks/src/test/scala/pl/touk/nussknacker/engine/benchmarks/spel/SpelSecurityBlacklistBenchmark.scala:23]: maybe move this above class definition

File: engine/benchmarks/src/test/scala/pl/touk/nussknacker/engine/benchmarks/spel/SpelSecurityBlacklistBenchmark.scala [hunk 2]
```diff
@@ -0,0 +1,20 @@
+package pl.touk.nussknacker.engine.benchmarks.spel
+
+import org.openjdk.jmh.annotations._
+
+import java.util.concurrent.TimeUnit
+
+
```
Comment (dswiecki) [inline engine/benchmarks/src/test/scala/pl/touk/nussknacker/engine/benchmarks/spel/SpelSecurityBlacklistBenchmark.scala:7]: try not introducing double empty lines

File: engine/interpreter/src/main/scala/pl/touk/nussknacker/engine/spel/internal/OptimizedEvaluationContext.scala [hunk 3]
```diff
@@ -32,11 +33,34 @@ class EvaluationContextPreparer(classLoader: ClassLoader,
     expressionImports.foreach(registerImport)
   }
 
+  private def blockBlacklisted(targetObject: scala.Any, methodName: String): Unit = {
+
+    val classFullName =
+      targetObject match {
+        case targetClass: Class[_] => targetClass.getName
```
Comment (dswiecki) [inline engine/interpreter/src/main/scala/pl/touk/nussknacker/engine/spel/internal/OptimizedEvaluationContext.scala:40]: I miss tests for this two cases, are we able to add them ?

File: engine/interpreter/src/main/scala/pl/touk/nussknacker/engine/spel/internal/OptimizedEvaluationContext.scala [hunk 4]
```diff
@@ -32,13 +37,36 @@ class EvaluationContextPreparer(classLoader: ClassLoader,
     expressionImports.foreach(registerImport)
   }
 
+  private val mockPackageBlackList: Set[Regex] = Set("^(java.lang.System).*$".r, "^(java.math).*$".r)
+
+  private def blockBlacklisted(targetObject: scala.Any, methodName: String): Unit = {
+    //todo: get blacklist from ExpressionConfig
+
+    val classFullName = targetObject.asInstanceOf[Class[_]].getName
+    mockPackageBlackList.find(blackListRegex =>
```
Comment (dswiecki) [inline engine/interpreter/src/main/scala/pl/touk/nussknacker/engine/spel/internal/OptimizedEvaluationContext.scala]: i'd move this logic to spelExpressionBlacklist - then you could add some unit tests and this logic would be separeted

**Comment by dswiecki** [comment]
and please rebase

File: engine/api/src/main/scala/pl/touk/nussknacker/engine/api/exception/BlacklistedPatternInvocationException.scala [hunk 5]
```diff
@@ -0,0 +1,7 @@
+package pl.touk.nussknacker.engine.api.exception
+
+import java.time.Instant
+
+case class BlacklistedPatternInvocationException(message: String,
+                                                 timestamp: Instant = Instant.now(), cause: Throwable = null)
```
Comment (dswiecki) [inline engine/api/src/main/scala/pl/touk/nussknacker/engine/api/exception/BlacklistedPatternInvocationException.scala:6]: You do not pass cause parameter to BlacklistedPatternInvocationException - probably can be just removed

File: engine/api/src/main/scala/pl/touk/nussknacker/engine/api/SpelExpressionBlacklist.scala [hunk 6]
```diff
@@ -0,0 +1,42 @@
...
+case class SpelExpressionBlacklist(blacklistedPatterns: Map[String, Regex]) {
+
+  private def getClassNameFromObject(targetObject: Object): String = {
+    targetObject match {
+      case targetClass: Class[_] => targetClass.getName
+      case _: Object => targetObject.getClass.getName
+    }
+  }
+
+  def blockBlacklisted(targetObject: Object, methodName: String): Unit = {
```
Comment (dswiecki) [inline engine/api/src/main/scala/pl/touk/nussknacker/engine/api/SpelExpressionBlacklist.scala:33]: I prefer when public methods are above private ones - but its just a note

**Review by dswiecki — APPROVED** [review]

File: engine/api/src/main/scala/pl/touk/nussknacker/engine/api/exception/BlacklistedPatternInvocationException.scala (hunk 5 above)
Comment (maciej-brzezinski) [reply engine/api/src/main/scala/pl/touk/nussknacker/engine/api/exception/BlacklistedPatternInvocationException.scala:6]: you are right, thanks

File: engine/api/src/main/scala/pl/touk/nussknacker/engine/api/SpelExpressionBlacklist.scala (hunk 6 above)
Comment (maciej-brzezinski) [reply engine/api/src/main/scala/pl/touk/nussknacker/engine/api/SpelExpressionBlacklist.scala:33]: Sounds like a good convention, I will try to implement it from now on, thanks

**Pull request merged**